stores/turbopuffer/src/vector/index.ts

```diff
@@ -0,0 +1,263 @@
+  type Vector,
+} from '@turbopuffer/turbopuffer';
+import { TurbopufferFilterTranslator } from './filter';
+
+export interface TurbopufferVectorOptions {
+  /** The API key to authenticate with. */
+  apiKey: string;
+  /** The base URL. Default is https://api.turbopuffer.com. */
+  baseUrl?: string;
+  /** The timeout to establish a connection, in ms. Default is 10_000. Only applicable in Node and Deno.*/
+  connectTimeout?: number;
+  /** The socket idle timeout, in ms. Default is 60_000. Only applicable in Node and Deno.*/
+  connectionIdleTimeout?: number;
+  /** The number of connections to open initially when creating a new client. Default is 0. */
+  warmConnections?: number;
+  /** Whether to compress requests and accept compressed responses. Default is true. */
+  compression?: boolean;
+  /**
+   * A callback function that takes an index name and returns a config object for that index.
+   * This allows you to define explicit schemas per index.
+   *
+   * Example:
+   * ```typescript
+   * schemaConfigForIndex: (indexName: string) => {
+   *   // Mastra's default embedding model and index for memory messages:
+   *   if (indexName === "memory_messages_384") {
+   *     return {
+   *       dimensions: 384,
+   *       schema: {
+   *         thread_id: {
+   *           type: "string",
+   *           filterable: true,
+   *         },
+   *       },
+   *     };
+   *   } else {
+   *     throw new Error(`TODO: add schema for index: ${indexName}`);
+   *   }
+   * },
+   * ```
+   */
+  schemaConfigForIndex?: (indexName: string) => {
+    dimensions: number;
+    schema: Schema;
+  };
+}
+
+export class TurbopufferVector extends MastraVector {
+  private client: Turbopuffer;
+  private filterTranslator: TurbopufferFilterTranslator;
+  // There is no explicit create index operation in Turbopuffer, so just register that
+  // someone has called createIndex() and verify that subsequent upsert calls are consistent
+  // with how the index was "created"
+  private createIndexCache: Map<
+    string,
+    {
+      dimension: number;
+      metric: 'cosine' | 'euclidean' | 'dotproduct';
+      tpufDistanceMetric: DistanceMetric;
+    }
+  > = new Map();
+  private opts: TurbopufferVectorOptions;
+
+  constructor(opts: TurbopufferVectorOptions) {
+    console.log(
+      `DEBUG turbopuffer constructor apiKey=${opts.apiKey ? '[REDACTED]' : undefined} baseUrl=${opts.baseUrl}`,
+    );
+    super();
+    this.filterTranslator = new TurbopufferFilterTranslator();
+    this.opts = opts;
+
+    const baseClient = new Turbopuffer(opts);
+    const telemetry = this.__getTelemetry();
+    this.client =
+      telemetry?.traceClass(baseClient, {
+        spanNamePrefix: 'turbopuffer-vector',
+        attributes: {
+          'vector.type': 'turbopuffer',
+        },
+      }) ?? baseClient;
+  }
+
+  async createIndex(
+    indexName: string,
+    dimension: number,
+    metric: 'cosine' | 'euclidean' | 'dotproduct' = 'cosine',
+  ): Promise<void> {
```

Comment:
createIndex is now in object params format, e.g.

`createIndex({
indexName, 
dimension, 
metric
}: {
indexName: string, 
dimension: number, 
metric: 'cosine' | 'euclidean' | 'dotproduct'
}
)
`